### DISCUSSION TITLE
Custom Typescript type for ScalarType

### DISCUSSION THREAD

**Discussion by hbriese** [edgedb/edgedb-js#593] [Ideas]
## Idea
Allow for custom typescript types to be used for scalars.
Scalars often have additional constraints, however this isn't reflected on the type level when interacting with them

### Example ScalarType
```
scalar type Address extending str {
   some constraints
}
```

Generates
```
export type $Address = $.ScalarType<"std::str", string>;
```

Ideally we could provide a custom typescript type rather than string (during generation)
```
export type $Address = $.ScalarType<"std::str", MyCustomType>;
```

Alternatively the generator could provide some constraints itself. There's likely a better way of doing this, but this is what I mean more concretely
```
export type $Address = $.ScalarType<"std::str", string & { isAddress: true } >;
```

**Comment by scotttrinh** [comment]
This feels similar to branding, for instance, here's [Zod's](https://zod.dev/?id=brand) implementation of that. Am I on the right track here with what you're proposing? Can you give some more motivation from the use-case perspective?
